File: huff-kt/src/main/kotlin/Heap.kt
// A heap implementing a priority queue.
// Break equal priorities by always choosing the node first added.

class Heap<T> {
    private class HeapNode<T>(var weight: Int, var serial: Int, var tree: T)

    // Here `size` is the number of active elements, `store` is as least as long as `size`
    // but may be longer, it is grown on demand.  Nodes are not shared outside the Heap and
    // may be reused.
    var size = 0
    private var store = Vector<HeapNode<T>>(0) {throw Exception("Bad")}
    private var serial = 0

    fun extractMax(): Pair<T, Int> {
        assert(size > 0)
        size--
        swap(0, size)
        val max = store[size]
        if (size > 1) {
            heapifyAtZero()
        }
        return Pair(max.tree, max.weight)
    }

    fun insert(weight: Int, tree: T) {
        push(weight, tree)
        var i = size-1
        var p = parent(i)
        while (i > 0 && greater(store[i], store[p])) {
            swap(i, p)
            i = p
            p = parent(i)
        }
    }

    private fun push(weight: Int, tree: T) {
        val loc = size
        size++
        if (loc >= store.size) {
            store.push(HeapNode<T>(weight, serial++, tree))
        } else {
            store[loc].weight = weight
            store[loc].serial = serial++
            store[loc].tree = tree
        }
    }

    private fun heapifyAtZero() {
        var loc = 0
        while (true) {
            var greatest = loc
            val l = left(loc)
            if (l < size && greater(store[l], store[loc])) {
                greatest = l
            }
            val r = right(loc)
            if (r < size && greater(store[r], store[greatest])) {
                greatest = r
            }
            if (greatest == loc) {
                break
            }
            swap(loc, greatest)
            loc = greatest
        }
    }

    private fun swap(i: Int, j: Int) {
        val t = store[i]
        store[i] = store[j]
        store[j] = t
    }

    private fun greater(a: HeapNode<T>, b: HeapNode<T>) = a.weight > b.weight || a.weight == b.weight && a.serial < b.serial

    private fun parent(loc: Int) = (loc - 1) / 2
    private fun left(loc: Int) = (loc * 2) + 1
    private fun right(loc: Int) = (loc + 1) * 2
}

File: huffrs/src/heap.rs
// A heap data structure, and utilities for using it as a priority queue.

use std::cmp;

struct Node<Val, Weight> {
    weight: Weight,
    val: Val
}

pub struct Heap<Val, Weight: Copy + cmp::PartialOrd> {
    xs: Vec<Node<Val, Weight>>
}

impl<Val, Weight: Copy + cmp::PartialOrd> Heap<Val, Weight> {
    pub fn new() -> Heap<Val, Weight> {
        Heap::<Val, Weight> {
            xs: Vec::<Node<Val, Weight>>::new()
        }
    }

    pub fn len(&self) -> usize {
        self.xs.len()
    }

    pub fn max_weight(&self) -> Weight {
        assert!(self.len() > 0);
        self.xs[0].weight   
    }

    pub fn extract_max(&mut self) -> (Val, Weight) {
        let l = self.xs.len();
        assert!(l > 0);
        self.xs.as_mut_slice().swap(0, l-1);
        let max = self.xs.pop().expect("POP");
        if l > 2 {
            self.heapify(0)
        }
        (max.val, max.weight)
    }

    pub fn insert(&mut self, weight: Weight, t: Val) {
        self.xs.push(Node{weight:weight, val:t});
        let mut i = self.xs.len() - 1;
        while i > 0 && self.xs[i].weight > self.xs[parent(i)].weight {
            self.xs.as_mut_slice().swap(i, parent(i));
            i = parent(i);
        }
    }

    fn heapify(&mut self, mut loc: usize) {
        loop {
            let mut greatest = loc;
            let l = left(loc);
            if l < self.xs.len() && self.xs[l].weight > self.xs[loc].weight {
                greatest = l;
            }
            let r = right(loc);
            if r < self.xs.len() && self.xs[r].weight > self.xs[greatest].weight {
                greatest = r;
            }
            if greatest == loc {
                break;
            }
            self.xs.as_mut_slice().swap(loc, greatest);
            loc = greatest;
        }
    }
}

fn parent(loc: usize) -> usize {
    (loc - 1) / 2
}

fn left(loc: usize) -> usize {
    (loc * 2) + 1
}

fn right(loc: usize) -> usize {
    (loc + 1) * 2
}